ignite/chainconfig/parser.go

```diff
@@ -0,0 +1,175 @@
+package chainconfig
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/imdario/mergo"
+	"gopkg.in/yaml.v2"
+
+	"github.com/ignite-hq/cli/ignite/chainconfig/common"
+	v0 "github.com/ignite-hq/cli/ignite/chainconfig/v0"
+	"github.com/ignite-hq/cli/ignite/pkg/xfilepath"
+)
+
+var (
+	// ConfigDirPath returns the path of configuration directory of Ignite.
+	ConfigDirPath = xfilepath.JoinFromHome(xfilepath.Path(".ignite"))
+
+	// ConfigFileNames is a list of recognized names as for Ignite's config file.
+	ConfigFileNames = []string{"config.yml", "config.yaml"}
+
+	DefaultVersion = "v0"
+
+	// Migration defines the version as the key and the config instance as the value
+	Migration = map[string]common.Config{"v0": &v0.ConfigYaml{}}
+)
+
+var (
+	// ErrCouldntLocateConfig returned when config.yml cannot be found in the source code.
+	ErrCouldntLocateConfig = errors.New(
+		"could not locate a config.yml in your chain. please follow the link for" +
+			"how-to: https://github.com/ignite-hq/cli/blob/develop/docs/configure/index.md")
+)
+
+// DefaultConf holds default configuration.
+var DefaultConf = v0.ConfigYaml{
+	Host: common.Host{
+		// when in Docker on MacOS, it only works with 0.0.0.0.
+		RPC:     "0.0.0.0:26657",
+		P2P:     "0.0.0.0:26656",
+		Prof:    "0.0.0.0:6060",
+		GRPC:    "0.0.0.0:9090",
+		GRPCWeb: "0.0.0.0:9091",
+		API:     "0.0.0.0:1317",
+	},
+	Build: common.Build{
+		Proto: common.Proto{
+			Path: "proto",
+			ThirdPartyPaths: []string{
+				"third_party/proto",
+				"proto_vendor",
+			},
+		},
+	},
+	Faucet: common.Faucet{
+		Host: "0.0.0.0:4500",
+	},
+}
+
+// Parse parses config.yml into UserConfig based on the version.
+func Parse(content []byte) (common.Config, error) {
+	// Read the version field
+	version, err := getConfigVersion(bytes.NewReader(content))
+	if err != nil {
+		return GetDefaultConfig(), err
+	}
+
+	conf := GetConfigInstance(version)
+	if err = yaml.NewDecoder(bytes.NewReader(content)).Decode(conf); err != nil {
+		return conf, err
+	}
+	if err = mergo.Merge(conf, conf.Default()); err != nil {
+		return GetDefaultConfig(), err
+	}
+	return conf, validate(conf)
+}
+
+// getConfigVersion returns the version in the io.Reader based on the field version.
+func getConfigVersion(r io.Reader) (string, error) {
+	var baseConf common.BaseConfigYaml
+	if err := yaml.NewDecoder(r).Decode(&baseConf); err != nil {
+		return DefaultVersion, err
+	}
+	if baseConf.Version != "" {
+		return baseConf.Version, nil
+	}
+	return DefaultVersion, nil
+}
+
+// GetConfigInstance retrieves correct config instance based on the version.
+func GetConfigInstance(version string) common.Config {
+	var config common.Config
+	var ok bool
+	if config, ok = Migration[version]; !ok {
+		// If there is no matching instance, return the config with the v0 version.
```

Comment:
if there is a version that's not supported, we should return an error to the user instead of using version 0

Reply:
still thinking...